File: chess_cpp/validators/PieceMoveValidator.cpp
//
// Created by Hrach Khachatryan on 12/15/19.
//

#include "PieceMoveValidator.h"

ValidationResponse *PieceMoveValidator::execute(Board* board, Spot* start, Spot* end) {
    bool valid = validate(board, start, end);
    if(nextValidator == nullptr || !valid){
        return new ValidationResponse(valid, !valid ? getNotValidMessage() : nullptr);
    }
    return nextValidator->execute(board, start, end);

}

void PieceMoveValidator::addValidator(std::shared_ptr<PieceMoveValidator> validator) {
    if (nextValidator == nullptr) {
        nextValidator = validator;
    } else {
        nextValidator->addValidator(validator);
    }
}

File: chess_java/src/chess/validators/PieceMoveValidator.java
package chess.validators;

import chess.Board;
import chess.Spot;
import chess.beans.ValidationResponse;

public abstract class PieceMoveValidator {

    private PieceMoveValidator nextValidator;


    public void addValidator(PieceMoveValidator validator) {
        if (nextValidator == null) {
            nextValidator = validator;
        } else {
            nextValidator.addValidator(validator);
        }
    }

    public ValidationResponse execute(Board board, Spot start, Spot end) {
        boolean valid = validate(board, start, end);
        if(nextValidator == null || !valid){
            return new ValidationResponse(valid, !valid ? getNotValidMessage() : null);
        }
        return nextValidator.execute(board, start, end);

    }

    public abstract boolean validate(Board board, Spot start, Spot end);

    protected abstract String getNotValidMessage();
}
